clash-prelude/src/Clash/Explicit/BlockRam/Model.hs

```diff
@@ -0,0 +1,315 @@
+  (MaybeX Bool, MaybeX Int, MaybeX writeEnable, a) ->
+  -- | Port B: enable, address, write enable, write data
+  (MaybeX Bool, MaybeX Int, MaybeX writeEnable, a) ->
+  -- | Updated memory, output value A, output value B
+  (Seq a, a, a)
+cycleBoth
+  SNat TdpbramModelConfig{..} prevA prevB ram0
+  (enAx, addrAx, byteEnaAx0, datA)
+  (enBx, addrBx, byteEnaBx0, datB) = (ram2, outA2, outB2)
+ where
+  conflict =
+    getConflict
+      (enAx, tdpIsActiveWriteEnable byteEnaAx1, addrAx)
+      (enBx, tdpIsActiveWriteEnable byteEnaBx1, addrBx)
+
+  writeWriteError = deepErrorX "conflicting write/write queries"
+  readWriteError = deepErrorX "conflicting read/write queries"
+
+  byteEnaAx1 = tdpMergeWriteEnable enAx byteEnaAx0
+  byteEnaBx1 = tdpMergeWriteEnable enBx byteEnaBx0
+
+  (datA1, datB1) = case conflict of
+    Just Conflict{cfWW=IsDefined True} -> (writeWriteError, writeWriteError)
+    Just Conflict{cfWW=IsX _} -> (writeWriteError, writeWriteError)
+    _ -> (datA, datB)
+
+  (wroteA, ram1) =
+    writeRam (SNat @nAddrs) tdpIsActiveWriteEnable tdpUpdateRam addrAx byteEnaAx1 datA1 ram0
+  (wroteB, ram2) =
+    writeRam (SNat @nAddrs) tdpIsActiveWriteEnable tdpUpdateRam addrBx byteEnaBx1 datB1 ram1
+
+  outA1 = case conflict of
+    Just Conflict{cfRWA=IsDefined True} -> readWriteError
+    Just Conflict{cfRWA=IsX _} -> readWriteError
+    _ -> fromMaybe (ram0 `Seq.index` MaybeX.unsafeToA addrAx) wroteA
+
+  outB1 = case conflict of
+    Just Conflict{cfRWB=IsDefined True} -> readWriteError
+    Just Conflict{cfRWB=IsX _} -> readWriteError
+    _ -> fromMaybe (ram0 `Seq.index` MaybeX.unsafeToA addrBx) wroteB
+
+  outA2 = if MaybeX.unsafeToA enAx then outA1 else prevA
+  outB2 = if MaybeX.unsafeToA enBx then outB1 else prevB
+
+-- | Write to a RAM and account for undefined values in the address, write enable,
+-- and data to write. Return read after write value.
+writeRam ::
+  forall nAddrs a writeEnable .
+  ( NFDataX a
+  , HasCallStack ) =>
+  SNat nAddrs ->
+  -- | Determine whether a write enable is active
+  (MaybeX writeEnable -> MaybeX Bool) ->
+  -- | Update memory with a defined address
+  (Int -> MaybeX writeEnable -> a -> Seq a -> Seq a) ->
+  -- | Address
+  MaybeX Int ->
+  -- | Byte enable
+  MaybeX writeEnable ->
+  -- | Data to write
+  a ->
+  -- | Memory to write to
+  Seq a ->
+  -- | (Read after write value, new memory)
+  (Maybe a, Seq a)
+writeRam SNat tdpIsActiveWriteEnable updateMem addrX byteEnableX dat mem
+  -- Read (do nothing)
+  | IsDefined False <- tdpIsActiveWriteEnable byteEnableX
+  = (Nothing, mem)
+
+  -- Undefined address and write enable or (partially) unknown
+  | IsX addrMsg <- addrX
+  , let msg = "Unknown address" <> "\nAddress error message: " <> addrMsg
+  = ( Just (deepErrorX msg)
+    , Seq.fromFunction (natToNum @nAddrs) (unknownAddr addrMsg) )
+
+  -- Write with defined address
+  | IsDefined addr <- addrX
+  = (Just dat, updateMem addr byteEnableX dat mem)
```

Comment:
I discovered an issue while working on the RTL test. How is this defined?

Given:
* memory contains a `(0 :: BitVector 16)` at that address
* We write with byte-enable `0b01` and data `0xAAAA`

Currently, memory is updated to contain `0x00AA` and the data output of the TDP block ram returns `0xAAAA` (which is the data input sans byte-enable masking), as can be easily seen in this line of code, `Just dat`.

I think "read after write" says we should output `0x00AA` on the data out as well, and in that case this code is wrong. I think we need to add a data-out to `tdpUpdateRam`.

Reply:
No wait, this makes more sense
```suggestion
  = let mem0 = updateMem addr byteEnableX dat mem
    in (Just $ mem0 `Seq.index` addr, mem0)
```

adjust for code style, but something equivalent to that.